File: packages/player/ios/PlaybackOrderManager.swift
import Foundation

/// Owns canonical queue occurrences and the playback order derived from them.
///
/// Occurrence tokens are independent from track metadata ids so duplicate media
/// entries remain distinct throughout a shuffle cycle.
final class PlaybackOrderManager {
    private var canonicalTokens: [Int] = []
    private var playbackOrderTokens: [Int] = []
    private var canonicalIndexByToken: [Int: Int] = [:]
    private var playbackPositionByToken: [Int: Int] = [:]
    private var nextToken = 0
    private(set) var isShuffleEnabled = false

    var size: Int {
        canonicalTokens.count
    }

    func reset(size: Int, currentIndex: Int = -1) {
        precondition(size >= 0, "Queue size must be non-negative")

        canonicalTokens = (0..<size).map { _ in
            defer { nextToken += 1 }
            return nextToken
        }
        reindexCanonical()
        if isShuffleEnabled {
            rebuildPlaybackOrder(currentIndex: currentIndex)
        } else {
            playbackOrderTokens.removeAll()
            playbackPositionByToken.removeAll()
        }
    }

    func insert(at index: Int, count: Int, currentIndex: Int = -1) {
        precondition(index >= 0 && index <= canonicalTokens.count, "Insert index is outside the canonical queue")
        precondition(count >= 0, "Insert count must be non-negative")
        guard count > 0 else { return }

        let newTokens = (0..<count).map { _ in
            defer { nextToken += 1 }
            return nextToken
        }
        canonicalTokens.insert(contentsOf: newTokens, at: index)
        reindexCanonical()

        guard isShuffleEnabled else { return }
        if playbackOrderTokens.isEmpty {
            rebuildPlaybackOrder(currentIndex: currentIndex)
            return
        }

        playbackOrderTokens.append(contentsOf: shuffled(newTokens))
        reindexPlaybackOrder()
    }

    func remove(indices: [Int]) {
        guard !indices.isEmpty else { return }

        let normalized = Array(Set(indices)).sorted(by: >)
        precondition(
            normalized.allSatisfy { canonicalTokens.indices.contains($0) },
            "Remove index is outside the canonical queue"
        )

        var removedTokens = Set<Int>()
        for index in normalized {
            removedTokens.insert(canonicalTokens.remove(at: index))
        }
        playbackOrderTokens.removeAll { removedTokens.contains($0) }
        reindexCanonical()
        reindexPlaybackOrder()
    }

    @discardableResult
    func setShuffleEnabled(_ enabled: Bool, currentIndex: Int) -> Bool {
        guard enabled != isShuffleEnabled else { return false }

        isShuffleEnabled = enabled
        if enabled {
            rebuildPlaybackOrder(currentIndex: currentIndex)
        } else {
            playbackOrderTokens.removeAll()
            playbackPositionByToken.removeAll()
        }
        return true
    }

    func nextIndex(after currentIndex: Int) -> Int? {
        guard isShuffleEnabled else {
            return currentIndex < canonicalTokens.count - 1 ? currentIndex + 1 : nil
        }

        guard let currentToken = canonicalTokens[safe: currentIndex] else {
            return firstIndex()
        }
        guard let position = playbackPositionByToken[currentToken], position < playbackOrderTokens.count - 1 else {
            return nil
        }
        return canonicalIndexByToken[playbackOrderTokens[position + 1]]
    }

    func previousIndex(before currentIndex: Int) -> Int? {
        guard isShuffleEnabled else {
            return currentIndex > 0 && currentIndex < canonicalTokens.count ? currentIndex - 1 : nil
        }

        guard
            let currentToken = canonicalTokens[safe: currentIndex],
            let position = playbackPositionByToken[currentToken],
            position > 0
        else {
            return nil
        }
        return canonicalIndexByToken[playbackOrderTokens[position - 1]]
    }

    func firstIndex() -> Int? {
        guard !canonicalTokens.isEmpty else { return nil }
        guard isShuffleEnabled else { return 0 }
        guard let firstToken = playbackOrderTokens.first else { return nil }
        return canonicalIndexByToken[firstToken]
    }

    func occurrenceToken(at index: Int) -> Int? {
        canonicalTokens[safe: index]
    }

    func canonicalIndex(for token: Int) -> Int? {
        canonicalIndexByToken[token]
    }

    private func rebuildPlaybackOrder(currentIndex: Int) {
        let effectiveCurrentIndex: Int
        if canonicalTokens.indices.contains(currentIndex) {
            effectiveCurrentIndex = currentIndex
        } else if canonicalTokens.isEmpty {
            effectiveCurrentIndex = -1
        } else {
            effectiveCurrentIndex = 0
        }

        let currentToken = canonicalTokens[safe: effectiveCurrentIndex]
        let remaining = canonicalTokens.filter { $0 != currentToken }
        playbackOrderTokens.removeAll(keepingCapacity: true)
        if let currentToken = currentToken {
            playbackOrderTokens.append(currentToken)
        }
        playbackOrderTokens.append(contentsOf: shuffled(remaining))
        reindexPlaybackOrder()
    }

    private func shuffled(_ tokens: [Int]) -> [Int] {
        guard tokens.count > 1 else { return tokens }
        var result = tokens
        for index in stride(from: result.count - 1, through: 1, by: -1) {
            result.swapAt(index, Int.random(in: 0...index))
        }
        return result
    }

    private func reindexCanonical() {
        canonicalIndexByToken.removeAll(keepingCapacity: true)
        for (index, token) in canonicalTokens.enumerated() {
            canonicalIndexByToken[token] = index
        }
    }

    private func reindexPlaybackOrder() {
        playbackPositionByToken.removeAll(keepingCapacity: true)
        for (index, token) in playbackOrderTokens.enumerated() {
            playbackPositionByToken[token] = index
        }
    }
}

private extension Collection {
    subscript(safe index: Index) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: packages/player/src/playback-order-manager.ts
type RandomSource = () => number;

function isCanonicalPermutation(order: readonly number[], size: number): boolean {
  if (order.length !== size) {
    return false;
  }
  const seen = new Set<number>();
  for (const index of order) {
    if (!Number.isInteger(index) || index < 0 || index >= size || seen.has(index)) {
      return false;
    }
    seen.add(index);
  }
  return true;
}

/**
 * Owns canonical queue occurrences and the playback order derived from them.
 *
 * Occurrence tokens are intentionally independent from TrackData.id: the same
 * media item may appear in a queue more than once and every occurrence must be
 * visited exactly once in a shuffle cycle.
 */
export class PlaybackOrderManager {
  private readonly random: RandomSource;
  private canonicalTokens: number[] = [];
  private playbackOrderTokens: number[] = [];
  private canonicalIndexByToken = new Map<number, number>();
  private playbackPositionByToken = new Map<number, number>();
  private nextToken = 0;
  private shuffleEnabled = false;

  constructor(random: RandomSource = Math.random) {
    this.random = random;
  }

  get isShuffleEnabled(): boolean {
    return this.shuffleEnabled;
  }

  get size(): number {
    return this.canonicalTokens.length;
  }

  reset(size: number, currentIndex = -1): void {
    if (!Number.isInteger(size) || size < 0) {
      throw new RangeError("Queue size must be a non-negative integer");
    }

    this.canonicalTokens = Array.from({ length: size }, () => this.nextToken++);
    this.reindexCanonical();
    if (this.shuffleEnabled) {
      this.rebuildPlaybackOrder(currentIndex);
    } else {
      this.playbackOrderTokens = [];
      this.playbackPositionByToken.clear();
    }
  }

  insert(index: number, count: number, currentIndex = -1): void {
    if (!Number.isInteger(index) || index < 0 || index > this.canonicalTokens.length) {
      throw new RangeError("Insert index is outside the canonical queue");
    }
    if (!Number.isInteger(count) || count < 0) {
      throw new RangeError("Insert count must be a non-negative integer");
    }
    if (count === 0) {
      return;
    }

    const newTokens = Array.from({ length: count }, () => this.nextToken++);
    this.canonicalTokens.splice(index, 0, ...newTokens);
    this.reindexCanonical();

    if (!this.shuffleEnabled) {
      return;
    }
    if (this.playbackOrderTokens.length === 0) {
      this.rebuildPlaybackOrder(currentIndex);
      return;
    }

    this.playbackOrderTokens.push(...this.shuffle(newTokens));
    this.reindexPlaybackOrder();
  }

  remove(indices: readonly number[]): void {
    if (indices.length === 0) {
      return;
    }

    const normalized = [...new Set(indices)].sort((a, b) => b - a);
    const invalidIndex = normalized.find(
      index => !Number.isInteger(index) || index < 0 || index >= this.canonicalTokens.length,
    );
    if (invalidIndex !== undefined) {
      throw new RangeError(`Remove index ${invalidIndex} is outside the canonical queue`);
    }

    const removedTokens = new Set<number>();
    for (const index of normalized) {
      removedTokens.add(this.canonicalTokens[index]!);
      this.canonicalTokens.splice(index, 1);
    }
    this.playbackOrderTokens = this.playbackOrderTokens.filter(token => !removedTokens.has(token));
    this.reindexCanonical();
    this.reindexPlaybackOrder();
  }

  setShuffleEnabled(enabled: boolean, currentIndex: number): boolean {
    if (enabled === this.shuffleEnabled) {
      return false;
    }

    this.shuffleEnabled = enabled;
    if (enabled) {
      this.rebuildPlaybackOrder(currentIndex);
    } else {
      this.playbackOrderTokens = [];
      this.playbackPositionByToken.clear();
    }
    return true;
  }

  getNextIndex(currentIndex: number): number {
    if (!this.shuffleEnabled) {
      return currentIndex < this.canonicalTokens.length - 1 ? currentIndex + 1 : -1;
    }

    const currentToken = this.canonicalTokens[currentIndex];
    if (currentToken === undefined) {
      return this.getFirstIndex();
    }
    const position = this.playbackPositionByToken.get(currentToken);
    if (position === undefined || position >= this.playbackOrderTokens.length - 1) {
      return -1;
    }
    return this.canonicalIndexByToken.get(this.playbackOrderTokens[position + 1]!) ?? -1;
  }

  getPreviousIndex(currentIndex: number): number {
    if (!this.shuffleEnabled) {
      return currentIndex > 0 && currentIndex < this.canonicalTokens.length ? currentIndex - 1 : -1;
    }

    const currentToken = this.canonicalTokens[currentIndex];
    if (currentToken === undefined) {
      return -1;
    }
    const position = this.playbackPositionByToken.get(currentToken);
    if (position === undefined || position <= 0) {
      return -1;
    }
    return this.canonicalIndexByToken.get(this.playbackOrderTokens[position - 1]!) ?? -1;
  }

  getFirstIndex(): number {
    if (this.canonicalTokens.length === 0) {
      return -1;
    }
    if (!this.shuffleEnabled) {
      return 0;
    }
    return this.canonicalIndexByToken.get(this.playbackOrderTokens[0]!) ?? -1;
  }

  /**
   * Canonical indices in playback order.
   *
   * When shuffle is disabled the playback order is the canonical order, so the
   * identity permutation is returned rather than an empty array.
   */
  getPlaybackOrder(): number[] {
    if (!this.shuffleEnabled) {
      return this.canonicalTokens.map((_, index) => index);
    }
    return this.playbackOrderTokens.map(token => this.canonicalIndexByToken.get(token) ?? -1);
  }

  /**
   * Restores a previously captured playback order.
   *
   * Rejects anything that is not a permutation of the current canonical indices, so a
   * stale persisted order cannot desynchronize traversal from the queue.
   */
  setPlaybackOrder(order: readonly number[]): boolean {
    if (!this.shuffleEnabled || !isCanonicalPermutation(order, this.canonicalTokens.length)) {
      return false;
    }

    this.playbackOrderTokens = order.map(index => this.canonicalTokens[index]!);
    this.reindexPlaybackOrder();
    return true;
  }

  getOccurrenceToken(index: number): number | undefined {
    return this.canonicalTokens[index];
  }

  getCanonicalIndex(token: number): number {
    return this.canonicalIndexByToken.get(token) ?? -1;
  }

  private rebuildPlaybackOrder(currentIndex: number): void {
    const effectiveCurrentIndex =
      currentIndex >= 0 && currentIndex < this.canonicalTokens.length
        ? currentIndex
        : this.canonicalTokens.length > 0
          ? 0
          : -1;
    const currentToken = this.canonicalTokens[effectiveCurrentIndex];
    const remaining =
      currentToken === undefined
        ? [...this.canonicalTokens]
        : this.canonicalTokens.filter(token => token !== currentToken);
    const shuffled = this.shuffle(remaining);
    this.playbackOrderTokens = currentToken === undefined ? shuffled : [currentToken, ...shuffled];
    this.reindexPlaybackOrder();
  }

  private shuffle(tokens: readonly number[]): number[] {
    const shuffled = [...tokens];
    for (let index = shuffled.length - 1; index > 0; index--) {
      const target = Math.floor(this.random() * (index + 1));
      [shuffled[index], shuffled[target]] = [shuffled[target]!, shuffled[index]!];
    }
    return shuffled;
  }

  private reindexCanonical(): void {
    this.canonicalIndexByToken.clear();
    this.canonicalTokens.forEach((token, index) => this.canonicalIndexByToken.set(token, index));
  }

  private reindexPlaybackOrder(): void {
    this.playbackPositionByToken.clear();
    this.playbackOrderTokens.forEach((token, index) => this.playbackPositionByToken.set(token, index));
  }
}
